### PR TITLE
Fix tests run on desktop

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Conditionalize tests appropriately so long path and new syntax
support are taken into account.

Move some test data into Theories for easier debugging.

Fix and add issue to colon check tests until we fix CoreCLR.

Note that there are still some failing tests, but this addresses all of the long path & invalid syntax issues (the majority).  #14986

@tarekgh @ianhays @danmosemsft

### CODE REVIEW

File: src/System.IO.FileSystem/tests/Tfm.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/tests/Tfm.cs]: is this  something useful enough to put in src\common\tests somewhere?

Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/Tfm.cs]: For this one I might add some to the enum in build tools. Some of them are pretty specific to my use case- figured I'd start here and follow up on making things more common.

File: src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs [hunk 2]
```diff
@@ -32,19 +32,15 @@ public void EmptyAsPath_ThrowsArgumentException()
             Assert.Throws<ArgumentException>(() => Create(string.Empty));
         }
 
-        [Fact]
-        public void PathWithInvalidCharactersAsPath_ThrowsArgumentException()
+        [Theory MemberData("PathsWithInvalidCharacters")]
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs]: Does this compile?

File: src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs [hunk 3]
```diff
@@ -130,7 +126,8 @@ public void ValidPathWithTrailingSlash()
             });
         }
 
-        [Fact]
+        [ConditionalFact("UsingNewNormalization")]
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs]: Nit: ```nameof(UsingNewNormalization)```
Could we also come up with a more descriptive name instead of "New"?  "New" will quickly lose its meaning.

File: src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs [hunk 4]
```diff
@@ -223,18 +220,17 @@ public void DirectoryWithComponentLongerThanMaxComponentAsPath_ThrowsPathTooLong
...
-            var paths = IOInputs.GetPathsWithInvalidColons();
-            Assert.All(paths, (path) =>
-            {
-                Assert.Throws<NotSupportedException>(() => Create(path));
-            });
+            Assert.Throws<ArgumentException>(() => Create(invalidPath));
         }
 
-        [Fact]
+        [ConditionalFact("AreAllLongPathsAvailable")]
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs]: Nit: nameof... same goes for all other such cases

File: src/System.IO.FileSystem/tests/Directory/Exists.cs [hunk 5]
```diff
@@ -53,17 +53,14 @@ public void ValidPathExists_ReturnsTrue()
             });
         }
 
-        [Fact]
-        public void PathWithInvalidCharactersAsPath_ReturnsFalse()
+        [Theory MemberData("PathsWithInvalidCharacters")]
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/Directory/Exists.cs]: Same question... this compiles?  This looks like a compiler bug that was recently fixed:
https://github.com/dotnet/roslyn/issues/14696
Please add a comma (here and elsewhere) so we're not dependent on the bug.
cc: @jaredpar

File: src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs [hunk 6]
```diff
@@ -219,9 +219,13 @@ public void WindowsSearchPatternLongSegment()
             Assert.Throws<PathTooLongException>(() => GetEntries(testDir.FullName, longName));
         }
 
-        [Fact]
+        [ConditionalFact("AreAllLongPathsAvailable")]
+        [SkipOnTargetFramework(Tfm.BelowNet462 | Tfm.Core50, "long path support added in 4.6.2")]
         public void SearchPatternLongPath()
         {
+            if (!PathFeatures.AreAllLongPathsAvailable())
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs]: Why is this needed if it's also the condition for the ConditionalFact?

File: src/System.IO.FileSystem/tests/PathFeatures.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace System.IO
+{
+    public static class PathFeatures
+    {
+        // Note that this class is using APIs that allow it to run on all platforms (including Core 5.0)
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: What is "Core 5.0"?

File: src/System.IO.FileSystem/tests/PathFeatures.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+        private static bool RunningOnCore
+        {
+            get
+            {
+                // Not particularly elegant
+                if (!s_onCore.HasValue)
+                    s_onCore = typeof(object).GetTypeInfo().Assembly.GetName().Name == "System.Private.CoreLib";
+
+                return s_onCore.Value;
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: This usage of bool? is not thread-safe.  It's possible two threads could use these properties in parallel and end up getting a wrong answer. (Same goes for other occurrences of this pattern.)

File: src/System.IO.FileSystem/tests/PathFeatures.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                {
+                    s_osEnabled = false;
+                }
+            }
+
+            return s_osEnabled.Value;
+        }
+
+        [DllImport("ntdll", ExactSpelling = true)]
+        static extern bool RtlAreLongPathsEnabled();
```
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: Nit: private

File: src/System.IO.FileSystem/tests/PathFeatures.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                return false;
+
+            Type t = typeof(object).GetTypeInfo().Assembly.GetType("System.AppContextSwitches");
+            var p = t.GetProperty(propertyName, BindingFlags.Static | BindingFlags.Public);
+
+            // If the switch actually exists use it, otherwise we predate the switch and are effectively on
+            return (bool)(p?.GetValue(null) ?? true);
+        }
+
+        private static bool RunningOnCore
```
Comment (tarekgh) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: I think we need a better way in general to detect that. System.Private.CoreLib exists in UAP too. so your check will include such platform. this maybe ok in your case.

File: src/System.IO.FileSystem/tests/Directory/Exists.cs (hunk 5 above)
Comment (jaredpar) [inline src/System.IO.FileSystem/tests/Directory/Exists.cs]: Yes please fix this.  The compiler will not accept this in 7.0

File: src/System.IO.FileSystem/tests/PathFeatures.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            {
+                // Not particularly elegant
+                if (!s_onCore.HasValue)
+                    s_onCore = typeof(object).GetTypeInfo().Assembly.GetName().Name == "System.Private.CoreLib";
+
+                return s_onCore.Value;
+            }
+        }
+
+        private static bool AreOsLongPathsEnabled()
```
Comment (tarekgh) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: could you please put assert for platform is Windows here.

**Comment by tarekgh** [comment]
other than the comments listed in this review, LGTM

File: src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs (hunk 2 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs]: Yes

File: src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs (hunk 3 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs]: I've tried to come up with a better name, but it comprises a lot of changes. Legacy is the term we shipped with the other way.

File: src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs (hunk 4 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs]: Ok, I'll fix that.

File: src/System.IO.FileSystem/tests/Directory/Exists.cs (hunk 5 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/Directory/Exists.cs]: Thanks for the heads up. Out of curiosity, why is it important to make commas mandatory?

File: src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs (hunk 6 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs]: Slipped through, thanks.

File: src/System.IO.FileSystem/tests/PathFeatures.cs (hunk 7 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: netcore50 variants.  If there is a better name I'm all for it.

File: src/System.IO.FileSystem/tests/PathFeatures.cs (hunk 8 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: How would that happen? Does the HasValue change state before the value is set? I don't really care that the value gets set multiple times.

File: src/System.IO.FileSystem/tests/PathFeatures.cs (hunk 11 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: Sure

File: src/System.IO.FileSystem/tests/PathFeatures.cs (hunk 9 above)
Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: Whoops, thanks

File: src/System.IO.FileSystem/tests/PathFeatures.cs (hunk 8 above)
Comment (stephentoub) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: > How would that happen? Does the HasValue change state before the value is set?

It's not written atomically.  So another thread could see HasValue == true even if Value is still the default.

Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: How is that possible if hasValue is set after the value is set? Is the problem with copying of the struct?

Comment (stephentoub) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: No fences are involved; it's completely valid for the compiler or the processor to reorder these reads/writes.

Comment (JeremyKuhne) [inline src/System.IO.FileSystem/tests/PathFeatures.cs]: Ah, I see. Thanks.

File: src/System.IO.FileSystem/tests/Directory/Exists.cs (hunk 5 above)
Comment (jaredpar) [inline src/System.IO.FileSystem/tests/Directory/Exists.cs]: It's the intended syntax of the language.  This was a bad omission in Roslyn 1.0 that we only recently caught.  The full impact, considering future features, hasn't ever been done for this.  Hence we moved back to the intended syntax.

**Comment by tarekgh** [comment]
@JeremyKuhne can you close #14986 now? or still there more need to be done?

**Comment by JeremyKuhne** [comment]
I'll close it as it addresses the specific issues they discuss. There are other failures still in this assembly.
